pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java

```diff
@@ -0,0 +1,360 @@
+  private static final String SCHEMA_PATH = "/SCHEMAS/";
+
+  private final String _clusterName;
+  private final String _controllerAddress;
+  private final String _brokerAddress;
+  private final ZKHelixAdmin _helixAdmin;
+  private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
+  private final Strategy _strategy;
+  private final Mode _mode;
+
+  private String _tableNamePattern = null;
+  private long _tableSizeThreshold = DEFAULT_TABLE_SIZE_THRESHOLD;
+  private long _cardinalityThreshold = DEFAULT_CARDINALITY_THRESHOLD;
+  private int _maxNumInvertedIndexAdded = DEFAULT_MAX_NUM_INVERTED_INDEX_ADDED;
+
+  public AutoAddInvertedIndex(@Nonnull String zkAddress, @Nonnull String clusterName, @Nonnull String controllerAddress,
+      @Nonnull String brokerAddress, @Nonnull Strategy strategy, @Nonnull Mode mode) {
+    _clusterName = clusterName;
+    _controllerAddress = controllerAddress;
+    _brokerAddress = brokerAddress;
+    _helixAdmin = new ZKHelixAdmin(zkAddress);
+    _propertyStore = new ZkHelixPropertyStore<>(zkAddress, new ZNRecordSerializer(),
+        PropertyPathConfig.getPath(PropertyType.PROPERTYSTORE, clusterName));
+    _strategy = strategy;
+    _mode = mode;
+  }
+
+  public void overrideDefaultSettings(@Nonnull String tableNamePattern, long tableSizeThreshold,
+      long cardinalityThreshold, int maxNumInvertedIndex) {
+    _tableNamePattern = tableNamePattern;
+    _tableSizeThreshold = tableSizeThreshold;
+    _cardinalityThreshold = cardinalityThreshold;
+    _maxNumInvertedIndexAdded = maxNumInvertedIndex;
+  }
+
+  public void run()
+      throws Exception {
+    if (_strategy == Strategy.QUERY) {
+      runQueryStrategy();
+    } else {
+      throw new IllegalStateException("Invalid Strategy: " + _strategy);
+    }
+  }
+
+  private void runQueryStrategy()
+      throws Exception {
+    // Get all resources in cluster
+    List<String> resourcesInCluster = _helixAdmin.getResourcesInCluster(_clusterName);
+
+    for (String tableName : resourcesInCluster) {
+      // Skip non-table resources
+      if (!tableName.endsWith("_OFFLINE") && !tableName.endsWith("_REALTIME")) {
+        continue;
+      }
+
+      // Skip tables that do not match the defined name pattern
+      if (_tableNamePattern != null && !tableName.matches(_tableNamePattern)) {
+        continue;
+      }
+      LOGGER.info("Table: {} matches the table name pattern: {}", tableName, _tableNamePattern);
+
+      // Get the inverted index config
+      AbstractTableConfig tableConfig = getTableConfig(tableName);
+      IndexingConfig indexingConfig = tableConfig.getIndexingConfig();
+      List<String> invertedIndexColumns = indexingConfig.getInvertedIndexColumns();
+      boolean autoGeneratedInvertedIndex = indexingConfig.isAutoGeneratedInvertedIndex();
+
+      // Handle auto-generated inverted index
+      if (autoGeneratedInvertedIndex) {
+        assert !invertedIndexColumns.isEmpty();
```

Comment:
Use Guava preconditions instead of Java asserts (Java asserts can be turned off)